### PR TITLE
Kamdelia and metrics mutex optimizations 

### PR DESCRIPTION
**Pull request by esadakar**
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/1861)

### CODE REVIEW

**Review by janos — APPROVED** [review]
This PR looks fine. I have one suggestion for some fo the future ones. It looks to me that some parts of Kademlia can be better separated into some new types, like waitNext, for clarity and in order to leverage type safety in method arguments for map keys and also to lock transparently and more reliably.

File: pkg/topology/kademlia/kademlia.go [hunk 1]
```diff
@@ -673,17 +673,18 @@ func (k *Kad) connect(ctx context.Context, peer swarm.Address, ma ma.Multiaddr)
...
 		var e *p2p.ConnectionBackoffError
 		failedAttempts := 0
 		if errors.As(err, &e) {
 			retryTime = e.TryAfter()
 		} else {
+			k.waitNextMu.Lock()
 			if info, ok := k.waitNext[peer.ByteString()]; ok {
 				failedAttempts = info.failedAttempts
 			}
 			failedAttempts++
```
Comment (aloknerurkar) [inline pkg/topology/kademlia/kademlia.go:686]: retryInfo is a struct and not a pointer. Incrementing `failedAttempts` here is not updating the struct in the map.

**Review by aloknerurkar — APPROVED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1861)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @acud, @anatollupacescu, and @mrekucci)

**Comment by Eknir** [comment]
To what degree do the changes in this PR fix the situation described in https://github.com/ethersphere/bee/issues/1860 ?

**Review by zelig — APPROVED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1861)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @acud, @anatollupacescu, and @mrekucci)

**Review by esadakar — COMMENTED** [review]
@janos I agree, it's messy atm.
@eknir It should resolve it.

> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1861#-:-MaXVd8u313zgBxSzJYK:b-s5voxt)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @acud, @anatollupacescu, and @mrekucci)

**Pull request merged**